File: ios/YamabikoChat/Data/Models/ChatMessageToolActivity.swift
import Foundation
import GRDB

struct ChatMessageToolActivity: Codable, FetchableRecord, MutablePersistableRecord, Equatable {
    static let databaseTableName = "chat_message_tool_activity"

    var id: Int64?
    var messageId: Int64?
    var variantId: Int64?
    var stepsJSON: String
    var providerTranscriptJSON: String?

    init(
        id: Int64? = nil,
        messageId: Int64? = nil,
        variantId: Int64? = nil,
        stepsJSON: String,
        providerTranscriptJSON: String? = nil
    ) {
        self.id = id
        self.messageId = messageId
        self.variantId = variantId
        self.stepsJSON = stepsJSON
        self.providerTranscriptJSON = providerTranscriptJSON
    }

    mutating func didInsert(_ inserted: InsertionSuccess) {
        id = inserted.rowID
    }

    var steps: [ToolActivityStep] {
        guard let data = stepsJSON.data(using: .utf8) else { return [] }
        return (try? JSONDecoder().decode([ToolActivityStep].self, from: data)) ?? []
    }

    var providerTranscript: [ProviderRequestMessage]? {
        guard let providerTranscriptJSON,
              let data = providerTranscriptJSON.data(using: .utf8)
        else { return nil }
        return try? JSONDecoder().decode([ProviderRequestMessage].self, from: data)
    }
}

struct ToolActivityStep: Codable, Sendable, Equatable, Identifiable {
    enum Status: String, Codable, Sendable {
        case running
        case completed
        case failed
    }

    var id: String
    var round: Int
    var toolName: String
    var title: String
    var detail: String
    var status: Status
    var resultCount: Int?
    var sources: [ToolSource]
    var errorMessage: String?
    var createdAtMs: Int64
    var resultPreview: String? = nil
    var outputPreview: String? = nil
    var artifactNames: [String]? = nil
    var durationMs: Int64? = nil

    var isWebActivity: Bool {
        toolName == WebSearchTool.name || toolName == FetchUrlTool.name
    }
}

struct ToolActivityEvent: Sendable, Equatable {
    enum Phase: Sendable, Equatable {
        case started
        case finished
    }

    var phase: Phase
    var call: ToolCall
    var result: ToolResult?
    var createdAtMs: Int64
}

struct ToolActivityPayload: Codable, Sendable, Equatable {
    var steps: [ToolActivityStep] = []
    var providerTranscript: [ProviderRequestMessage] = []
    var attachmentPaths: [String] = []

    init(
        steps: [ToolActivityStep] = [],
        providerTranscript: [ProviderRequestMessage] = [],
        attachmentPaths: [String] = []
    ) {
        self.steps = steps
        self.providerTranscript = providerTranscript
        self.attachmentPaths = attachmentPaths
    }

    private enum CodingKeys: String, CodingKey {
        case steps, providerTranscript, attachmentPaths
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        steps = try container.decodeIfPresent([ToolActivityStep].self, forKey: .steps) ?? []
        providerTranscript = try container.decodeIfPresent([ProviderRequestMessage].self, forKey: .providerTranscript) ?? []
        attachmentPaths = try container.decodeIfPresent([String].self, forKey: .attachmentPaths) ?? []
    }

    mutating func apply(_ event: ToolActivityEvent) {
        var step = Self.step(for: event)
        if let index = steps.firstIndex(where: { $0.id == step.id }) {
            step.round = steps[index].round
            steps[index] = step
        } else {
            step.round = (steps.map(\.round).max() ?? 0) + 1
            steps.append(step)
        }
        steps.sort { $0.round < $1.round }

        guard event.phase == .finished, let result = event.result else { return }
        for path in result.artifacts.map(\.path) where !attachmentPaths.contains(path) {
            attachmentPaths.append(path)
        }
        providerTranscript.removeAll { message in
            message.toolCallId == event.call.id || message.toolCalls?.contains(where: { $0.id == event.call.id }) == true
        }
        providerTranscript.append(
            ProviderRequestMessage(role: "assistant", content: "", toolCalls: [event.call])
        )
        providerTranscript.append(
            ProviderRequestMessage(
                role: "tool",
                content: result.content,
                toolCallId: event.call.id,
                toolName: event.call.name,
                toolResultIsError: result.isError
            )
        )
    }

    mutating func failRunning(message: String) {
        for index in steps.indices where steps[index].status == .running {
            steps[index].status = .failed
            steps[index].errorMessage = message
        }
    }

    private static func step(for event: ToolActivityEvent) -> ToolActivityStep {
        let arguments = jsonObject(event.call.argumentsJSON)
        let query = (arguments?["query"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
        let goal = (arguments?["goal"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
        let rawURL = arguments?["url"] as? String
        let host = rawURL.flatMap { URL(string: $0)?.host }
        let isSearch = event.call.name == WebSearchTool.name
        let isFetch = event.call.name == FetchUrlTool.name
        let isPython = event.call.name == PythonExecuteTool.name
        let code = (arguments?["code"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
        let title = isPython
            ? L10n.text("Pythonを実行")
            : (isSearch
                ? L10n.text("Webを検索")
                : (isFetch ? L10n.text("ページを確認") : displayName(event.call.name)))
        let detail = isPython
            ? (code?.trimmedNonEmpty ?? L10n.text("コードを確認中"))
            : (isSearch
                ? (query?.trimmedNonEmpty ?? L10n.text("検索語を確認中"))
                : (isFetch
                    ? [host, goal?.trimmedNonEmpty].compactMap { $0 }.joined(separator: " — ").trimmedNonEmpty ?? L10n.text("ページを確認中")
                    : genericArgumentSummary(arguments, raw: event.call.argumentsJSON)))
        let result = event.result
        let resultObject = result.flatMap { jsonObject($0.content) }
        let resultCount = (resultObject?["results"] as? [Any])?.count
        let nestedError = (resultObject?["error"] as? [String: Any])?["message"] as? String
        let error = result?.isError == true
            ? (nestedError?.trimmedNonEmpty ?? (resultObject?["error"] as? String)?.trimmedNonEmpty ?? L10n.text("ツールの実行に失敗しました"))
            : nil
        let sources = deduplicatedSources(result?.sources ?? [])
        let stdout = (resultObject?["stdout"] as? String)?.trimmedNonEmpty
        let stderr = (resultObject?["stderr"] as? String)?.trimmedNonEmpty
        let outputPreview = [stdout, stderr].compactMap { $0 }.joined(separator: "\n").trimmedNonEmpty
        let resultPreview = (resultObject?["result_repr"] as? String)?.trimmedNonEmpty
        let artifactNames = result?.artifacts.map(\.name).filter { !$0.isEmpty }
        let durationMs = (resultObject?["duration_ms"] as? NSNumber)?.int64Value
        return ToolActivityStep(
            id: event.call.id,
            round: 0,
            toolName: event.call.name,
            title: title,
            detail: detail,
            status: event.phase == .started ? .running : (result?.isError == true ? .failed : .completed),
            resultCount: resultCount,
            sources: sources,
            errorMessage: error,
            createdAtMs: event.createdAtMs,
            resultPreview: resultPreview,
            outputPreview: outputPreview,
            artifactNames: artifactNames?.isEmpty == false ? artifactNames : nil,
            durationMs: durationMs
        )
    }

    private static func displayName(_ name: String) -> String {
        name.replacingOccurrences(of: "_", with: " ")
    }

    private static func genericArgumentSummary(_ object: [String: Any]?, raw: String) -> String {
        guard let object else { return raw }
        let values = object.keys.sorted().compactMap { key -> String? in
            switch object[key] {
            case let value as String:
                return "\(key): \(value)"
            case let value as NSNumber:
                return "\(key): \(value)"
            default:
                return nil
            }
        }
        return values.joined(separator: " · ").trimmedNonEmpty ?? raw
    }

    private static func jsonObject(_ value: String) -> [String: Any]? {
        guard let data = value.data(using: .utf8) else { return nil }
        return try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    }

    private static func deduplicatedSources(_ sources: [ToolSource]) -> [ToolSource] {
        var seen: Set<String> = []
        return sources.filter { seen.insert($0.url).inserted }
    }
}

File: app/src/main/java/com/porarri/yamabikochat/data/local/ChatMessageToolActivity.kt
package com.porarri.yamabikochat.data.local

import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import com.porarri.yamabikochat.data.model.ProviderRequestMessage
import com.porarri.yamabikochat.data.model.ToolActivityEvent
import com.porarri.yamabikochat.data.model.ToolSource
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.net.URI
import java.io.File

@Serializable
data class ToolActivityStep(
    val id: String,
    val round: Int = 0,
    val toolName: String = "",
    val title: String = "",
    val detail: String = "",
    val status: Status = Status.completed,
    val resultCount: Int? = null,
    val sources: List<ToolSource> = emptyList(),
    val errorMessage: String? = null,
    val createdAtMs: Long = System.currentTimeMillis(),
    val artifactNames: List<String>? = null
) {
    @Serializable
    enum class Status {
        running,
        completed,
        failed
    }

    companion object {
        private val json = Json { ignoreUnknownKeys = true; isLenient = true; encodeDefaults = true }

        fun decodeSteps(raw: String): List<ToolActivityStep> =
            runCatching { json.decodeFromString<List<ToolActivityStep>>(raw) }.getOrDefault(emptyList())

        fun encodeSteps(steps: List<ToolActivityStep>): String =
            json.encodeToString(steps)
    }
}

@Entity(
    tableName = "chat_message_tool_activity",
    foreignKeys = [
        ForeignKey(
            entity = ChatMessage::class,
            parentColumns = ["id"],
            childColumns = ["messageId"],
            onDelete = ForeignKey.CASCADE
        ),
        ForeignKey(
            entity = ChatMessageVariant::class,
            parentColumns = ["id"],
            childColumns = ["variantId"],
            onDelete = ForeignKey.CASCADE
        )
    ],
    indices = [
        Index(value = ["messageId"]),
        Index(value = ["variantId"])
    ]
)
data class ChatMessageToolActivity(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val messageId: Long? = null,
    val variantId: Long? = null,
    val stepsJSON: String,
    val providerTranscriptJSON: String? = null,
    val attachmentPathsJSON: String? = null
) {
    val steps: List<ToolActivityStep>
        get() = ToolActivityStep.decodeSteps(stepsJSON)

    val providerTranscript: List<ProviderRequestMessage>?
        get() = providerTranscriptJSON?.let { encoded ->
            runCatching { transcriptCodec.decodeFromString<List<ProviderRequestMessage>>(encoded) }.getOrNull()
        }

    val attachmentPaths: List<String>
        get() = attachmentPathsJSON?.let { encoded ->
            runCatching { transcriptCodec.decodeFromString<List<String>>(encoded) }.getOrNull()
        }.orEmpty()

    companion object {
        private val transcriptCodec = Json {
            ignoreUnknownKeys = true
            isLenient = true
            encodeDefaults = true
        }

        fun encodeProviderTranscript(contents: List<ProviderRequestMessage>): String =
            transcriptCodec.encodeToString(contents)

        fun encodeAttachmentPaths(paths: List<String>): String = transcriptCodec.encodeToString(paths)
    }
}

@Serializable
data class ToolActivityPayload(
    val steps: List<ToolActivityStep> = emptyList(),
    val providerTranscript: List<ProviderRequestMessage> = emptyList(),
    val attachmentPaths: List<String> = emptyList()
) {
    fun applying(event: ToolActivityEvent): ToolActivityPayload {
        val previous = steps.firstOrNull { it.id == event.call.id }
        val updatedStep = stepFor(event).copy(
            round = previous?.round ?: ((steps.maxOfOrNull { it.round } ?: 0) + 1)
        )
        var updatedSteps = (steps.filterNot { it.id == updatedStep.id } + updatedStep)
            .sortedBy { it.round }
        if (event.phase != ToolActivityEvent.Phase.finished || event.result == null) {
            return copy(steps = updatedSteps)
        }
        val callId = event.call.id
        var updatedAttachments = attachmentPaths
        if (event.result.artifacts.isNotEmpty()) {
            val logicalNames = event.result.artifacts.mapTo(mutableSetOf()) { it.name }
            event.result.artifacts.forEach { artifact ->
                updatedAttachments = updatedAttachments.filterNot { isPersistedVersionOf(it, artifact.name) } + artifact.path
            }
            updatedSteps = updatedSteps.map { step ->
                if (step.id == callId) {
                    step.copy(artifactNames = event.result.artifacts.map { it.name })
                } else {
                    step.copy(artifactNames = step.artifactNames?.filterNot(logicalNames::contains)?.takeIf { it.isNotEmpty() })
                }
            }
        }
        val transcript = providerTranscript.filterNot { message ->
            message.toolCallId == callId || message.toolCalls?.any { it.id == callId } == true
        } + listOf(
            ProviderRequestMessage(role = "assistant", content = "", toolCalls = listOf(event.call)),
            ProviderRequestMessage(
                role = "tool",
                content = event.result.content,
                toolCallId = callId,
                toolName = event.call.name,
                toolResultIsError = event.result.isError
            )
        )
        return copy(steps = updatedSteps, providerTranscript = transcript, attachmentPaths = updatedAttachments)
    }

    fun failRunning(message: String): ToolActivityPayload = copy(
        steps = steps.map {
            if (it.status == ToolActivityStep.Status.running) {
                it.copy(status = ToolActivityStep.Status.failed, errorMessage = message)
            } else it
        }
    )

    private fun stepFor(event: ToolActivityEvent): ToolActivityStep {
        val arguments = runCatching { payloadJson.parseToJsonElement(event.call.argumentsJSON).jsonObject }.getOrNull()
        val isSearch = event.call.name == "web_search"
        val isEditor = event.call.name == "str_replace_editor"
        val query = arguments?.get("query")?.jsonPrimitive?.contentOrNull?.trim().orEmpty()
        val goal = arguments?.get("goal")?.jsonPrimitive?.contentOrNull?.trim().orEmpty()
        val host = arguments?.get("url")?.jsonPrimitive?.contentOrNull?.takeIf { it.isNotBlank() }
            ?.let { runCatching { URI(it).host }.getOrNull() }
        val editorCommand = arguments?.get("command")?.jsonPrimitive?.contentOrNull.orEmpty()
        val editorPath = arguments?.get("path")?.jsonPrimitive?.contentOrNull.orEmpty()
        val detail = if (isEditor) listOf(editorCommand, editorPath).filter { it.isNotBlank() }.joinToString(" — ")
        else if (isSearch) query.ifBlank { "検索語を確認中" }
        else listOfNotNull(host, goal.takeIf { it.isNotBlank() }).joinToString(" — ").ifBlank { "ページを確認中" }
        val resultObject = event.result?.let { runCatching { payloadJson.parseToJsonElement(it.content).jsonObject }.getOrNull() }
        val resultCount = resultObject?.get("results")?.let { runCatching { it.jsonArray.size }.getOrNull() }
        val error = if (event.result?.isError == true) {
            resultObject?.get("error")?.jsonPrimitive?.contentOrNull?.takeIf { it.isNotBlank() }
                ?: "ツールの実行に失敗しました"
        } else null
        return ToolActivityStep(
            id = event.call.id,
            toolName = event.call.name,
            title = if (isEditor) "ファイルを編集" else if (isSearch) "Webを検索" else "ページを確認",
            detail = detail,
            status = if (event.phase == ToolActivityEvent.Phase.started) ToolActivityStep.Status.running
                else if (event.result?.isError == true) ToolActivityStep.Status.failed else ToolActivityStep.Status.completed,
            resultCount = resultCount,
            sources = event.result?.sources.orEmpty().distinctBy { it.url },
            errorMessage = error,
            createdAtMs = event.createdAtMs,
            artifactNames = event.result?.artifacts?.map { it.name }?.takeIf { it.isNotEmpty() }
        )
    }

    companion object {
        private val payloadJson = Json { ignoreUnknownKeys = true; isLenient = true }

        private fun isPersistedVersionOf(path: String, logicalName: String): Boolean {
            val candidate = File(path).name
            if (candidate == logicalName) return true
            val source = File(logicalName)
            val extension = source.extension.takeIf { it.isNotEmpty() }?.let { ".${Regex.escape(it)}" }.orEmpty()
            val stem = if (source.extension.isEmpty()) source.name else source.nameWithoutExtension
            return candidate.matches(Regex("${Regex.escape(stem)} \\(\\d+\\)$extension"))
        }
    }
}
